Kernel/Arch/x86_64/Interrupts.cpp

```diff
@@ -397,6 +397,7 @@ void unregister_generic_interrupt_handler(u8 interrupt_number, GenericInterruptH
         if (!shared_handler->sharing_devices_count()) {
             handler_slot = nullptr;
             revert_to_unused_handler(interrupt_number);
+            delete shared_handler;
```

Comment:
This seems flimsy. Can we change this code to using `OwnPtr` instead of having a `new` and `delete` far apart from each other?

Reply:
It definitely is flimsy :D For e.g, we allocate memory as follows:
```
        handler_slot = nullptr;
        SharedIRQHandler::initialize(interrupt_number);
        VERIFY(handler_slot);
```
I don't see how we can use `OwnPtr` here as the s_interrupt_handler is a pointer of arrays. The same issues exist with the `UnhandledIRQHandler` where we sprinkle new and delete.

So probably the solution is to do a complete overhaul of s_interrupt_handler to be `static OwnPtr<GenericInterruptHandler> s_interrupt_handler` ?

Reply:
> This seems flimsy. Can we change this code to using `OwnPtr` instead of having a `new` and `delete` far apart from each other?

This is not the correct solution. Many IRQ handlers are also ref-counted in many places (I could give examples, but it's easy to find them so let's assume you can believe me).
I tried to do this about 2 years ago and got to this very conclusion that this is not the right path. And yes, raw pointers are not cool either, but this whole code is very raw and very low-level, so it will need a different solution.

Reply:
> If you ask me, I'd suggest to never unregister a shared IRQ handler - if we got to the point we need, we might as well just leave it there forever. This will make this delete not needed, as well as it will simplify the code too.
I know that the chances of shrared IRQ Handler not having any handlers is less, but how is leaving that memory not cleaned the right solution?

Reply:
> > If you ask me, I'd suggest to never unregister a shared IRQ handler - if we got to the point we need, we might as well just leave it there forever. This will make this delete not needed, as well as it will simplify the code too.
> 
> I know that the chances of shrared IRQ Handler not having any handlers is less, but how is leaving that memory not cleaned the right solution?

This is not about leaving that memory not cleaned up - it's about that you never revert it to unused handler.

Reply:
> If you ask me, I'd suggest to never unregister a shared IRQ handler - if we got to the point we need, we might as well just leave it there forever. This will make this delete not needed, as well as it will simplify the code too.

Ah, got it. So if we register a new IRQ in that spot, the same SharedIRQHandler can be used. 

I will keep the changes as is for this PR but once we do a complete overhaul of how we hook the interrupt handler, we can do things differently.

Reply:
> This is not the correct solution. Many IRQ handlers are also ref-counted in many places (I could give examples, but it's easy to find them so let's assume you can believe me).

For IRQ handlers that are ref-counted, they would be arriving here via `unref()` -> `will_be_destroyed()` -> `unregister_interrupt_handler()` -> `unregister_generic_interrupt_handler()`

If we `delete` them here in `unregister_generic_interrupt_handler()` as the patch does, that means that `unref()` will then `delete` the same object a second time. That can't be right.

Reply:
Hmm, got it. For now I will remove this patch. I think the way we store function callbacks in `s_interrupt_handler` needs a bit of love. I will remove the patch that does a `delete` and probaby create a new series with your suggestion to use smart pointers.